staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -360,3 +382,72 @@ func (s *WatchServer) HandleWS(ws *websocket.Conn) {
 		}
 	}
 }
+
+func recordWatchListLatency(ctx context.Context, reqScope *RequestScope, reqOpts *metainternalversion.ListOptions) {
+	requestReceivedTimestamp, ok := request.ReceivedTimestampFrom(ctx)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unable to measure watchlist latency because no received ts found in the ctx, gvr: %s", reqScope.Resource))
+		return
+	}
+	scope := determineRequestScope(ctx, reqOpts)
+	elapsedSeconds := time.Now().Sub(requestReceivedTimestamp).Seconds()
+	gvr := reqScope.Resource
+
+	metrics.WatchListLatencies.WithContext(ctx).WithLabelValues(gvr.Group, gvr.Version, gvr.Resource, scope).Observe(elapsedSeconds)
+}
+
+func shouldRecordWatchListLatency(event watch.Event) bool {
+	if event.Type != watch.Bookmark || !utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		return false
+	}
+	// as of today the initial-events-end annotation is added only to a single event
+	// by the watch cache and only when certain conditions are met
+	//
+	// for more please read https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list
+	hasAnnotation, err := storage.HasInitialEventsEndBookmarkAnnotation(event.Object)
+	if err != nil {
+		utilruntime.HandleError(fmt.Errorf("unable to determine if the obj has the required annotation for measuring watchlist latency, obj %T: %v", event.Object, err))
+		return false
+	}
+	return hasAnnotation
+}
+
+// determineRequestScope computes the request scope
+//
+// note that normally we would use RequestInfo for computation
+// bur we cannot, due to the same reasons mentioned in determineRequestNamespaceAndName
+func determineRequestScope(ctx context.Context, opts *metainternalversion.ListOptions) string {
```

Comment:
We should move this function to pkg/endpoint/metrics/metrics.go [there is CleanScope there as an example]

Reply:
Maybe let's call it `CleanListScope` or sth like that....

Reply:
We could do that. So I assume you would also like to move the `determineRequestNamespaceAndName` function to the same place, right ?

Reply:
moved, ptal.